File: sync_headers_footers.py
import os
import re

# Use current directory instead of hardcoded path
base_dir = os.path.dirname(os.path.abspath(__file__))
header_file = os.path.join(base_dir, "includes", "header-template.html")
footer_file = os.path.join(base_dir, "includes", "footer-template.html")

if not os.path.exists(header_file) or not os.path.exists(footer_file):
    print(f"Error: Header or footer template not found in {os.path.join(base_dir, 'includes')}")
    exit(1)

with open(header_file, 'r', encoding='utf-8') as f:
    header_template = f.read()

with open(footer_file, 'r', encoding='utf-8') as f:
    footer_template = f.read()

def get_active_link(filename):
    if filename == 'index.html': return '/'
    clean = '/' + filename.replace('.html', '')
    if clean == '/local-seo' or clean == '/local-citations':
        return '/services'
    return clean

# The standard head tags that should be in every file
HEAD_REQUIREMENTS = [
    '<link rel="preconnect" href="https://fonts.googleapis.com">',
    '<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>',
    '<link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&family=Outfit:wght@500;600;700&display=swap" rel="stylesheet">',
    '<link rel="stylesheet" href="css/styles.css">',
    '<link rel="icon" type="image/png" href="assets/favicon.png">'
]

total_files = 0
updated_files = 0

for root, dirs, files in os.walk(base_dir):
    # Skip some directories
    if any(skip in root for skip in ['.git', 'node_modules', 'includes', 'assets', 'css', 'js', '.agent']):
        continue
        
    for file in files:
        if file.endswith('.html') and file != '404.html':
            total_files += 1
            file_path = os.path.join(root, file)
            
            try:
                with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
                    content = f.read()
            except Exception as e:
                print(f"Error reading {file_path}: {e}")
                continue
                
            original_content = content
            
            # 1. REMOVE EXISTING HEADER AND FOOTER (Clean slate)
            header_pattern = re.compile(r'(?s)<!--.*?STANDARD HEADER TEMPLATE.*?-->\s*<header.*?</header>', re.DOTALL)
            content = header_pattern.sub('', content)
            content = re.sub(r'(?s)<header class="header" id="header".*?</header>', '', content)
            
            footer_pattern = re.compile(r'(?s)<!--.*?STANDARD FOOTER TEMPLATE.*?-->\s*<footer.*?</footer>', re.DOTALL)
            content = footer_pattern.sub('', content)
            content = re.sub(r'(?s)<footer class="footer".*?</footer>', '', content)

            # 2. FIX STRUCTURAL TAGS (Repair mode)
            # Ensure </head> and <body> exist. If they were accidentally deleted or never there.
            if '</head>' not in content or '<body>' not in content:
                # Find the best place to close head and start body
                # After title or last meta/link
                last_head_idx = -1
                for marker in ['</title>', '</meta>', '<meta', '<link', '</script>']:
                    idx = content.rfind(marker)
                    if idx != -1:
                        end_idx = content.find('>', idx) + 1
                        if end_idx > last_head_idx:
                            last_head_idx = end_idx
                
                if last_head_idx != -1:
                    content = content[:last_head_idx] + "\n</head>\n\n<body>" + content[last_head_idx:]
                else:
                    # Fallback: find first <main> or first section
                    main_idx = content.find('<main')
                    if main_idx != -1:
                        content = content[:main_idx] + "</head>\n\n<body>\n    " + content[main_idx:]

            # 3. ENSURE HEAD REQUIREMENTS (CSS, Fonts)
            head_end = content.find('</head>')
            if head_end != -1:
                head_part = content[:head_end]
                missing_tags = []
                for req in HEAD_REQUIREMENTS:
                    # Check for parts of the requirement to avoid duplicates
                    req_id = req.split('href="')[1].split('"')[0] if 'href="' in req else req
                    if req_id not in head_part:
                        missing_tags.append(req)
                
                if missing_tags:
                    content = content[:head_end] + "    " + "\n    ".join(missing_tags) + "\n" + content[head_end:]

            # 4. PREPARE TEMPLATES
            active_href = get_active_link(file)
            current_header = header_template
            current_footer = footer_template
            
            # Update 'active' class in header
            current_header = current_header.replace('class="nav__link active"', 'class="nav__link"')
            if active_href:
                target_str = f'href="{active_href}" class="nav__link"'
                current_header = current_header.replace(target_str, f'href="{active_href}" class="nav__link active"')

            # 5. INJECT HEADER (After <body>)
            body_start = content.find('<body')
            if body_start != -1:
                body_tag_end = content.find('>', body_start) + 1
                content = content[:body_tag_end] + "\n    " + current_header + content[body_tag_end:]

            # 6. INJECT FOOTER (Before </body>)
            # First ensure </main> exists if <main> exists
            if '<main' in content and '</main>' not in content:
                # Find start of scripts or </body>
                insertion_point = content.find('<script')
                if insertion_point == -1:
                    insertion_point = content.find('</body>')
                
                if insertion_point != -1:
                    content = content[:insertion_point] + "\n    </main>" + content[insertion_point:]

            body_end = content.find('</body>')
            if body_end != -1:
                content = content[:body_end] + "\n    " + current_footer + content[body_end:]

            # 7. SAVE IF CHANGED
            if content.strip() != original_content.strip():
                with open(file_path, 'w', encoding='utf-8') as f:
                    f.write(content)
                updated_files += 1

print(f"Scanned {total_files} HTML files.")
print(f"Updated header/footer and repaired structure in {updated_files} files.")

File: sync_headers_footers.js
const fs = require('fs');
const path = require('path');

const baseDir = __dirname;
const headerFile = path.join(baseDir, "includes", "header-template.html");
const footerFile = path.join(baseDir, "includes", "footer-template.html");

if (!fs.existsSync(headerFile) || !fs.existsSync(footerFile)) {
    console.error(`Error: Header or footer template not found in ${path.join(baseDir, 'includes')}`);
    process.exit(1);
}

const headerTemplate = fs.readFileSync(headerFile, 'utf8');
const footerTemplate = fs.readFileSync(footerFile, 'utf8');

function getActiveLink(filename) {
    if (filename === 'index.html') return '/';
    const clean = '/' + filename.replace('.html', '');
    if (clean === '/local-seo' || clean === '/local-citations') {
        return '/services';
    }
    if (clean === '/founder' || clean === '/team') {
        return '/about';
    }
    return clean;
}

const HEAD_REQUIREMENTS = [
    '<link rel="preconnect" href="https://fonts.googleapis.com">',
    '<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>',
    '<link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&family=Outfit:wght@500;600;700&display=swap" rel="stylesheet">',
    '<link rel="stylesheet" href="/css/styles.css">',
    '<link rel="icon" type="image/png" href="/assets/favicon.png">'
];

const skipDirs = ['.git', 'node_modules', 'includes', 'assets', 'css', 'js', '.agent', 'temp-repo'];

function walk(dir, callback) {
    const files = fs.readdirSync(dir);
    for (const file of files) {
        const filePath = path.join(dir, file);
        const stat = fs.statSync(filePath);
        if (stat.isDirectory()) {
            if (skipDirs.includes(file)) continue;
            walk(filePath, callback);
        } else {
            callback(filePath, file);
        }
    }
}

let totalFiles = 0;
let updatedFiles = 0;

walk(baseDir, (filePath, file) => {
    if (!file.endsWith('.html') || file === '404.html') return;
    totalFiles++;

    let content = fs.readFileSync(filePath, 'utf8');
    const originalContent = content;

    // 1. REMOVE EXISTING HEADER AND FOOTER (Clean slate - safe separate replacement)
    content = content.replace(/<!--\s*STANDARD HEADER TEMPLATE\s*-->/g, '');
    content = content.replace(/<header class="header" id="header"[\s\S]*?<\/header>/g, '');

    content = content.replace(/<!--\s*STANDARD FOOTER TEMPLATE\s*-->/g, '');
    content = content.replace(/<footer class="footer"[\s\S]*?<\/footer>/g, '');

    // Remove existing back-to-top buttons to avoid duplication
    content = content.replace(/<button id="back-to-top"[\s\S]*?<\/button>/gi, '');

    // Remove existing WhatsApp floating buttons to avoid duplication
    content = content.replace(/<a[^>]*class="whatsapp-btn"[\s\S]*?<\/a>/gi, '');

    // 2. FIX STRUCTURAL TAGS (Ensure </head> and <body> exist)
    if (!content.includes('</head>') || !content.includes('<body>')) {
        let lastHeadIdx = -1;
        const markers = ['</title>', '</meta>', '<meta', '<link', '</script>'];
        for (const marker of markers) {
            const idx = content.lastIndexOf(marker);
            if (idx !== -1) {
                const endIdx = content.indexOf('>', idx) + 1;
                if (endIdx > lastHeadIdx) {
                    lastHeadIdx = endIdx;
                }
            }
        }
        if (lastHeadIdx !== -1) {
            content = content.slice(0, lastHeadIdx) + "\n</head>\n\n<body>" + content.slice(lastHeadIdx);
        } else {
            const mainIdx = content.indexOf('<main');
            if (mainIdx !== -1) {
                content = content.slice(0, mainIdx) + "</head>\n\n<body>\n    " + content.slice(mainIdx);
            }
        }
    }

    // 3. ENSURE HEAD REQUIREMENTS (CSS, Fonts)
    const headEnd = content.indexOf('</head>');
    if (headEnd !== -1) {
        const headPart = content.slice(0, headEnd);
        const missingTags = [];
        for (const req of HEAD_REQUIREMENTS) {
            const hrefMatch = req.match(/href="([^"]+)"/);
            const reqId = hrefMatch ? hrefMatch[1] : req;
            if (!headPart.includes(reqId)) {
                missingTags.push(req);
            }
        }
        if (missingTags.length > 0) {
            content = content.slice(0, headEnd) + "    " + missingTags.join('\n    ') + "\n" + content.slice(headEnd);
        }
    }

    // 4. PREPARE TEMPLATES
    const activeHref = getActiveLink(file);
    let currentHeader = headerTemplate;
    let currentFooter = footerTemplate;

    // Update active class in header
    currentHeader = currentHeader.replace('class="nav__link active"', 'class="nav__link"');
    if (activeHref) {
        const targetStr = `href="${activeHref}" class="nav__link"`;
        currentHeader = currentHeader.replace(targetStr, `href="${activeHref}" class="nav__link active"`);
    }

    // 5. INJECT HEADER (After <body>)
    const bodyStart = content.indexOf('<body');
    if (bodyStart !== -1) {
        const bodyTagEnd = content.indexOf('>', bodyStart) + 1;
        content = content.slice(0, bodyTagEnd) + "\n    " + currentHeader + content.slice(bodyTagEnd);
    }

    // 6. INJECT FOOTER (Before </body>)
    if (content.includes('<main') && !content.includes('</main>')) {
        let insertionPoint = content.indexOf('<script');
        if (insertionPoint === -1) {
            insertionPoint = content.indexOf('</body>');
        }
        if (insertionPoint !== -1) {
            content = content.slice(0, insertionPoint) + "\n    </main>" + content.slice(insertionPoint);
        }
    }

    const bodyEnd = content.indexOf('</body>');
    if (bodyEnd !== -1) {
        content = content.slice(0, bodyEnd) + "\n    " + currentFooter + content.slice(bodyEnd);
    }

    // 7. SAVE IF CHANGED
    if (content.trim() !== originalContent.trim()) {
        fs.writeFileSync(filePath, content, 'utf8');
        updatedFiles++;
        console.log(`Updated header/footer in ${filePath}`);
    }
});

console.log(`Scanned ${totalFiles} HTML files.`);
console.log(`Updated header/footer and repaired structure in ${updatedFiles} files.`);
